service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/UploadAdditionalApplicationsController.java

```diff
@@ -0,0 +1,179 @@
+import uk.gov.hmcts.reform.fpl.events.FailedPBAPaymentEvent;
+import uk.gov.hmcts.reform.fpl.model.CaseData;
+import uk.gov.hmcts.reform.fpl.model.PBAPayment;
+import uk.gov.hmcts.reform.fpl.model.common.AdditionalApplicationsBundle;
+import uk.gov.hmcts.reform.fpl.model.common.C2DocumentBundle;
+import uk.gov.hmcts.reform.fpl.model.common.Element;
+import uk.gov.hmcts.reform.fpl.model.common.OtherApplicationsBundle;
+import uk.gov.hmcts.reform.fpl.service.PbaNumberService;
+import uk.gov.hmcts.reform.fpl.service.additionalapplications.ApplicationsFeeCalculator;
+import uk.gov.hmcts.reform.fpl.service.additionalapplications.UploadAdditionalApplicationsService;
+import uk.gov.hmcts.reform.fpl.service.payment.PaymentService;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.apache.commons.lang3.ObjectUtils.defaultIfNull;
+import static uk.gov.hmcts.reform.fpl.enums.ApplicationType.C2_APPLICATION;
+import static uk.gov.hmcts.reform.fpl.enums.YesNo.NO;
+import static uk.gov.hmcts.reform.fpl.enums.YesNo.YES;
+import static uk.gov.hmcts.reform.fpl.utils.CaseDetailsHelper.removeTemporaryFields;
+import static uk.gov.hmcts.reform.fpl.utils.ElementUtils.element;
+
+@Api
+@Slf4j
+@RestController
+@RequestMapping("/callback/upload-additional-applications")
+@RequiredArgsConstructor(onConstructor = @__(@Autowired))
+public class UploadAdditionalApplicationsController extends CallbackController {
+    private static final String DISPLAY_AMOUNT_TO_PAY = "displayAmountToPay";
+    private static final String AMOUNT_TO_PAY = "amountToPay";
+    private static final String TEMPORARY_C2_DOCUMENT = "temporaryC2Document";
+    private static final String TEMPORARY_OTHER_APPLICATIONS_BUNDLE = "temporaryOtherApplicationsBundle";
+    private final PaymentService paymentService;
+    private final PbaNumberService pbaNumberService;
+    private final UploadAdditionalApplicationsService uploadC2DocumentsService;
+    private final ApplicationsFeeCalculator applicationsFeeCalculator;
+
+    @PostMapping("/get-fee/mid-event")
+    public AboutToStartOrSubmitCallbackResponse handleMidEvent(@RequestBody CallbackRequest callbackRequest) {
+        CaseDetails caseDetails = callbackRequest.getCaseDetails();
+        CaseData caseData = getCaseData(caseDetails);
+
+        caseDetails.getData().putAll(applicationsFeeCalculator.calculateFee(caseData));
+
+        return respond(caseDetails);
+    }
+
+    @PostMapping("/validate/mid-event")
+    public AboutToStartOrSubmitCallbackResponse handleValidateMidEvent(@RequestBody CallbackRequest callbackRequest) {
+        CaseDetails caseDetails = callbackRequest.getCaseDetails();
+        CaseData caseData = getCaseData(caseDetails);
+
+        List<String> errors = new ArrayList<>();
+
+        if (caseData.getTemporaryPbaPayment() != null) {
+            var updatedPbaNumber = pbaNumberService.update(caseData.getTemporaryPbaPayment().getPbaNumber());
```

Comment:
maybe pbaNumberService should accept PBAPayment and returns updated PBAPayment

Reply:
added method to take PBAPayment as a parameter. I left the existing method as it is still used in uploadC2 event